### PR TITLE
Draft: start working on editor undo/redo feature

### PR DESCRIPTION
**Pull request by archimede67**
Suggested by Pulsar yet again, I think this feature could be highly useful for a lot of mappers, so I'm trying to implement a functional history system. I'm planing to make most of the actions that can be done inside the editor as reversible. This includes:
- add/remove layer
- add/remove group
- ~~rename layer/group~~
- change tile (placing, removing)
- edit quad (color, position, center, rotation)
- edit layer properties (group, color, position, ...)
- edit group properties (order, para x/y, ...)
- enveloppes
	- name
	- points (value, time)
	- checkbox "Synchronized"
	- all editboxes
	- undo delete (if possible)
- commands
	- adding
	- deleting
	- editing
- ~~add image (if possible)~~
- ~~add sound (if possible)~~
- change order of layer/group
- change orientation value (editbox)

Pulsar also suggested to make a popup to warn users about some actions that cannot be undone. For example if a user wants to automap a layer, there will first be a popup with a warning text "Warning, this action cannot be reversed.". There will be an "OK" button but also a checkbox to not show the popup next time. There will also be a client command to disable the popups.

Right now, two buttons, "Undo" and "Redo" have been added in the toolbar, with shortcuts (respectively ctrl+z and ctrl+y). I also started working on some actions to undo which are changing color of a tile layer and a quad vertex.

image

The current limit of the history is 50 actions. This will probably be configurable through a command.

There is still a lot to be done and this is still a work in progress. If you have some ideas to add to this or some suggestion on what else can be undone, or if you think some things don't have to be reversed, please let me know. Also, please keep in mind that the code might change a lot and that the first commits might not be the better ones.

## Todo

- [x] Add "Undo" and "Redo" buttons with shortcuts
- [x] Undo: "add/remove layer"
- [x] Undo: "add/remove group"
- [x] Undo: "change tile (placing, removing)"
- [x] Undo: "edit quad"
	- [x] color
	- [x] position
	- [x] pivot
	- [x] rotation
	- [x] square, align & aspect ratio operations
	- [x] points properties (uv, position, ...)
- [ ] Undo: "edit layer properties"
	- [ ] group
	- [ ] order
	- [x] color
	- [ ] width/height
	- [ ] image
	- [ ] color env
	- [ ] color to
- [x] Undo: "edit group properties"
	- [x] order
	- [x] pos X/pos Y
	- [x] para X/para Y
	- [x] clip X/clip Y/clip W/clip H
- [ ] Undo: "edit enveloppes"
	- [ ] name
	- [ ] points (value, time)
	- [ ] checkbox "Synchronized"
	- [ ] all editboxes
	- [ ] undo delete (if possible)
- [x] Undo: "edit commands"
	- [x] adding
	- [x] deleting
	- [x] editing
	- [x] moving
- [ ] Undo: "change orientation value (editbox)"

## Checklist
- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by Pulsartw** [comment]
Nice to see this being worked on!

I have some suggestions. One that should be an obvious improvement:
* envelopes and server settings should have their own history. These 2 are not tied to map editing in any way, so any action there would just pollute global history. I would suggest that the window which is open takes precedence, and if none are open it operates on the global history.

And a couple of others:
* renaming of any kind could be ignored
* assets could be ignored
* live automapping could be easily reversed, so might as well add it
* some configurable memory usage limit

As for the things that should have a popup, I can only think of appending and automapping right now.

**Comment by Chairn** [comment]
Wow, nice work in progress there. I think it would be better if you separate undo code from the editor in a different file. editor code is already quite big enough.

**Comment by archimede67** [comment]
> Wow, nice work in progress there. I think it would be better if you separate undo code from the editor in a different file. editor code is already quite big enough.

I eventually will separate a lot of code into different files, especially all the different `CEditorAction`s. But for simplicity reasons I'm keeping them in editor code for now. Some parts cannot really be separated from editor code though.

**Comment by def-** [comment]
Oh wow, really cool that you're working on this @archimede67 . I assume you're doing it in a safe way instead of just having an extra thread that reads into the random memory while editor is running? (which caused crashes)

**Comment by archimede67** [comment]
> Oh wow, really cool that you're working on this @archimede67 . I assume you're doing it in a safe way instead of just having an extra thread that reads into the random memory while editor is running? (which caused crashes)

I'm recording all the possible actions that a user can do, there's no thread involved. Each action knows the old and new "state" so it's possible to revert back to the previous state, before the action was performed or redo the action. There might still be crashes if all actions are not handled properly, which already happend but I should have fixed them.

Of course once I think I'm done I would love to see people "beta-test" the feature :)

File: src/game/editor/layer_tiles.cpp [hunk 1]
```diff
@@ -64,7 +64,7 @@ CLayerTiles::CLayerTiles(const CLayerTiles &Other) :
 	m_Switch = Other.m_Switch;
 	m_Tune = Other.m_Tune;
 
-	mem_copy(m_aFileName, Other.m_aFileName, IO_MAX_PATH_LENGTH);
+	str_copy(m_aFileName, Other.m_aFileName, sizeof(m_aFileName));
```
Comment (Kaffeine) [inline src/game/editor/layer_tiles.cpp:67]:
```suggestion
	str_copy(m_aFileName, Other.m_aFileName);
```

File: src/game/editor/layer_tiles.cpp [hunk 2]
```diff
@@ -912,10 +912,22 @@ int CLayerTiles::RenderProperties(CUIRect *pToolBox)
 	}
 	else if(Prop == PROP_COLOR)
 	{
+		CColor PrevColor = m_Color;
 		m_Color.r = (NewVal >> 24) & 0xff;
 		m_Color.g = (NewVal >> 16) & 0xff;
 		m_Color.b = (NewVal >> 8) & 0xff;
 		m_Color.a = NewVal & 0xff;
+
+		if(!(PrevColor == m_Color))
```
Comment (Kaffeine) [inline src/game/editor/layer_tiles.cpp:921]:
```suggestion
		if(PrevColor != m_Color)
```

File: src/game/editor/layer_tiles.cpp [hunk 3]
```diff
@@ -1098,11 +1129,27 @@ CLayerTele::CLayerTele(int w, int h) :
 	mem_zero(m_pTeleTile, (size_t)w * h * sizeof(CTeleTile));
 }
 
+CLayerTele::CLayerTele(const CLayerTele &Other) :
+	CLayerTiles(Other)
+{
+	str_copy(m_aName, "Tele", sizeof(m_aName));
```
Comment (Kaffeine) [inline src/game/editor/layer_tiles.cpp:1135]:
```suggestion
	str_copy(m_aName, "Tele");
```

File: src/game/editor/layer_tiles.cpp [hunk 4]
```diff
@@ -1327,6 +1374,19 @@ CLayerSpeedup::CLayerSpeedup(int w, int h) :
 	mem_zero(m_pSpeedupTile, (size_t)w * h * sizeof(CSpeedupTile));
 }
 
+CLayerSpeedup::CLayerSpeedup(const CLayerSpeedup &Other) :
+	CLayerTiles(Other)
+{
+	str_copy(m_aName, "Speedup", sizeof(m_aName));
```
Comment (Kaffeine) [inline src/game/editor/layer_tiles.cpp:1380]:
```suggestion
	str_copy(m_aName, "Speedup");
```

Comment (archimede67) [reply src/game/editor/layer_tiles.cpp:1380]: Does it really hurt to be explicit about the size?

Comment (Kaffeine) [reply src/game/editor/layer_tiles.cpp:1380]: It hurts the code readability.

Comment (heinrich5991) [reply src/game/editor/layer_tiles.cpp:1380]: > Does it really hurt to be explicit about the size?

When you omit the size, it's inserted by the compiler. The argument is that you cannot make size mistakes that way. It's not the same as the `strcpy` function which doesn't do any bounds checks at all.

Comment (archimede67) [reply src/game/editor/layer_tiles.cpp:1380]: Alright, I will do

**Comment by heinrich5991** [comment]
What's the status of this @archimede67? :)

**Comment by VoxelDoesCode** [comment]
Hi! Just updated the UI a bit in #6029 . When you continue on this, do you think it's possible to use these symbols for undo/redo?
UNDO: `\xEF\x8B\xAA`
REDO: `\xEF\x8B\xB9`

image

**Comment by heinrich5991** [comment]
Status: No new changes since last status check. @archimede67 Still interested in this in general? It's okay to say "yes" even if you don't have time to spend on this right now.

**Comment by archimede67** [comment]
@heinrich5991 Yes, I am still interested in that feature, and as you've said, right now I just don't have a time to work on it

**Comment by heinrich5991** [comment]
Superseded by #7393, I guess.

**Pull request closed without merging**